app/src/main/java/org/mozilla/focus/settings/ManualAddSearchEngineSettingsFragment.java

```diff
@@ -65,18 +82,101 @@ public boolean onOptionsItemSelected(MenuItem item) {
                 final ManualAddSearchEnginePreference pref = (ManualAddSearchEnginePreference) findPreference(getString(R.string.pref_key_manual_add_search_engine));
                 final boolean engineValid = pref.validateEngineNameAndShowError(engineName);
                 final boolean searchValid = pref.validateSearchQueryAndShowError(searchQuery);
+                final boolean isPartialSuccess = engineValid && searchValid;
 
-                final boolean isSuccess = engineValid && searchValid;
-                if (isSuccess) {
-                    final SharedPreferences sharedPreferences = getSearchEngineSharedPreferences();
-                    SearchEngineManager.addSearchEngine(sharedPreferences, getActivity(), engineName, searchQuery);
-                    Snackbar.make(rootView, R.string.search_add_confirmation, Snackbar.LENGTH_SHORT).show();
-                    getFragmentManager().popBackStack();
+                if (isPartialSuccess) {
+                    // TODO: add spinner, disable button until async task completes.
+                    // Hide the keyboard because:
+                    // - It's awkward to show the keyboard while waiting for a response
+                    // - We want it hidden when we return to the previous screen (on success)
+                    // - An expanded keyboard hides the success snackbar
+                    ViewUtils.hideKeyboard(rootView);
+                    new ValidateSearchEngineAsyncTask(this, engineName, searchQuery).execute();
```

Comment:
Can you cancel this task in the fragment if the fragment gets destroyed? This should be a short-running task, and requires user action to trigger, but it might still be weird if the user 1) rotates the device, and then 2) goes back to the search engines. Settings doesn't handle rotation configChanges, so probably want to have an onCancelled that *doesn't* add the search engine.

Reply:
Also, optionally check for network connectivity first.

Reply:
Added a commit.